Second prototype version:  Ban inference tensor in autograd

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51274 Second prototype version:  Ban inference tensor in autograd**

Differential Revision: [D26438984](https://our.internmc.facebook.com/intern/diff/D26438984)

Commit:
initial changes.

[ghstack-poisoned]

Commit:
Update on "initial changes."




[ghstack-poisoned]

Commit:
Update on "initial changes."




[ghstack-poisoned]

Commit:
Update on "First prototype version: fails at erroring out in the following case."


torch::Tensor a = torch::ones({1, 2, 3});
  {
    torch::AutoInferenceMode guard(true);

    benchmark("view_inference", count, [&]() {
      b = a.view(-1);
    });
    // this should be banned: if a has an existing autograd session this might result in silent error.
    a.add_(1);
}

[ghstack-poisoned]

Commit:
Update on "First prototype version: fails at erroring out in the following case."


torch::Tensor a = torch::ones({1, 2, 3});
  {
    torch::AutoInferenceMode guard(true);

    benchmark("view_inference", count, [&]() {
      b = a.view(-1);
    });
    // this should be banned: if a has an existing autograd session this might result in silent error.
    a.add_(1);
}

[ghstack-poisoned]

bhosmer commented on c10/core/TensorImpl.h:
```diff
@@ -1503,6 +1508,18 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+    if (!track_view_ && track_view) {
+      TORCH_CHECK(false, "Changing inference_only tensor to track_view=True is not allowd");
+    } 
+    version_counter_ = VariableVersion(track_view ? 0 : -1);
```
Aside: should we care about what `version_counter_` might be in the `track_view` true case? I.e., would this ever zero out the version counter when we would want to preserve it?

Not to sidetrack from the main feature decision, just a note for if we go this way.

Commit:
Update on "First prototype version: fails at erroring out in the following case."

```
torch::Tensor a = torch::ones({1, 2, 3});
  {
    torch::AutoInferenceMode guard(true);

    benchmark("view_inference", count, [&]() {
      b = a.view(-1);
    });
    // this should be banned: if a has an existing autograd session this might result in silent error.
    a.add_(1);
}
```

[ghstack-poisoned]

Commit:
Update on "Second prototype version:  Ban inference tensor in autograd"




[ghstack-poisoned]
